internal/reload/watcher_test.go

```diff
@@ -0,0 +1,91 @@
+package reload
+
+import (
+	"bufio"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/quickfeed/quickfeed/internal/rand"
+)
+
+func TestWatcher(t *testing.T) {
+	dir := t.TempDir()
+	filename := filepath.Join(dir, rand.String())
+
+	// create a file to modify later
+	// to trigger WRITE events in the watcher
+	file, err := os.Create(filename)
+	if err != nil {
+		t.Errorf("failed to create file: %v", err)
+	}
+	defer file.Close()
+
+	watcher, err := NewWatcher(dir)
+	if err != nil {
+		t.Fatalf("failed to create watcher: %v", err)
+	}
+
+	// create a server with a handler the client can connect to
+	mux := http.NewServeMux()
+	mux.HandleFunc("/watch", watcher.Handler)
+	server := httptest.NewServer(mux)
+	defer server.Close()
+
+	go func() {
+		// wait some time before writing to the file
+		time.Sleep(2 * time.Second)
+		// try to write to the file multiple times
+		// in case the server is slow to start
+		for i := 0; i < 3; i++ {
+			// write to the file to trigger the watcher
+			// to send an event to the client
+			_, err = file.WriteString(rand.String())
+			if err != nil {
+				t.Errorf("failed to write to file: %v", err)
+			}
+		}
+	}()
+
+	// connect to the server
+	resp, err := server.Client().Get(server.URL + "/watch")
+	if err != nil {
+		t.Fatalf("failed to get from server: %v", err)
+	}
+	defer resp.Body.Close()
+
+	// continously read from the response body
+	eventChan := make(chan string, 1)
+	go func() {
+		reader := bufio.NewReader(resp.Body)
+		for {
+			line, err := reader.ReadString('\n')
+			if err != nil {
+				break
+			}
+			// ignore empty lines; not all received messages contain data
+			if line != "" {
+				// send the message to the event channel
+				eventChan <- line
+				break
```

Comment:
Is this intended? It will read only one line and exit the for-loop/goroutine.

Reply:
We're only interested in checking if we receive a non-empty message.
Since we only check for a single event below there was no point in sending multiple events on the channel here.

The main thing tested here is that the file modification on line 47 is picked up and sent as expected to our client.
I trigger a write 3 times, with some delay between writes to account for startup.

The received event will be identical for all three events ("data: $filename").